cpp/ql/src/experimental/Security/CWE/CWE-297/ValidationCertificateHostMismatch.ql

```diff
@@ -0,0 +1,73 @@
+/**
+ * @name Validation certificate host mismatch.
+ * @description Lack of certificate name validation allows any valid certificate to be used. And that can lead to disclosure.
+ * @kind problem
+ * @id cpp/validation-certificate-host-mismatch
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-297
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+predicate checkHostnameOlder1() {
```

Comment:
Reading this predicate out loud, this is what I'm reading:

This predicate holds if there exists a function call _anywhere_ in the program that calls `X509_get_ext_d2i`, and either:
- There exists a use of the macro `sk_GENERAL_NAME_num` or a use of the macro `sk_GENERAL_NAME_value`, or
- There exists a call to the function `sk_GENERAL_NAME_value` or a call to the function `sk_GENERAL_NAME_num`.

That doesn't make a lot of sense to me. Can `sk_GENERAL_NAME_num` and `sk_GENERAL_NAME_value` really both be macros and function calls (maybe depending on versions?). Also: Shouldn't this predicate have any parameters? Right now, this is just a property of the _entire_ codebase that either holds of doesn't hold (and I can make it hold by adding a random call to `X509_get_ext_d2i`, `sk_GENERAL_NAME_num` and `sk_GENERAL_NAME_value` anywhere in the program.

What are you really trying to model with this predicate?

Reply:
sorry for the delay in replies.

in this predicate I have omitted call connectivity control due to the presence of macros, in the rest of the predicates I use `GVN` for call connectivity control.